fixed mysql dialect to use `options` instead of `encoding` and hardcoded values

sio4 opened the pull request:
With this PR and previously merged PR, all hardcoded options for MySQL dialect are removed and it can be configured with `options` in `database.yml`. For example,

```yaml
development:
  dialect: mysql
  options:
    collation: utf8mb4_general_ci
    readTimeout: 1s
    parseTime: true
    multiStatements: true
```

Previously,

* `collation`: was configured with `encoding` key (high level key but only used by MySQL)
* `readTimeout`: was hardcoded for generated URL and changed from time to time.
* `parseTime` and `multiStatements`: was hardcoded for generated URL.

For `encoding` configuration, deprecated message was added to notice users.

There are some minor bugfixes related to these changes and added some test cases to cover most of the codes.

(related to #281)

u007 commented:
hi Sio, 
it converts configurations to mysql url right?
im okay with this, better control within database.yaml

stanislas-m commented:
`parseTime=true` is a mandatory option for pop, and `multiStatements: true` is required for the migrations to work. Is it still enforced?

sio4 commented:
Hi @stanislas-m,

> `parseTime=true` is a mandatory option for pop (so we can work with timestamp fields), and `multiStatements: true` is required for the migrations to work. Is it still enforced?

I added two maps in `finalizerMySQL()`. One is for *default options* and those values will be applied automatically and silently when Finalize() is called, but only when the user-defined values not exist. I believe users should have the right to control all thing. Another one is for *forced options* as you said, but currently, I just wrote it works as same as defaults. The only different thing is that if the option/value is not found on the configuration, just put a warning message.

I think most users do not set that values manually. so It will be (semi-)forced by default. but if a user specified that options with spending their time, maybe there are some reasons for them. My opinion is, user's choice is always the first even though the system is not working by that. Same reason, if the user uses URL, more complicated way, it just passed as is.

It was a problem because there is no way to configure some details. So the user should use URL if they need some special setup, and it makes problems of not set the parameter you mentioned. Anyway, now users can just use yaml configuration hierarchy, just with their interesting parameter like `readTimeout` for slow systems, so I think it is enough.

Well, I also concerned about it, as commented on that part. :-)

sio4 commented:
See the test cases. it shows all:

`Test_MySQL_Finalizer_Default_Options()`

is showing it fills default values if the user does not specified parameters.

`Test_MySQL_Finalizer_Preserve_User_Defined_Options()`

is showing it just preserve user-defined parameters.

stanislas-m commented:
Also, I understand your point of view for the default params. It makes sense, even though I'm sure some people will blame the product because it allowed an incompatible config. But we'll manage!

sio4 commented:
Thanks for your understanding. Yes, I agree with your concerns. Some people may complain if their configuration is not working properly. We need to manage it and should write more detailed documents. If my English is good enough, I want to write some more on the pages but mine is so bad. :-(

> Also, I understand your point of view for the default params. It makes sense, even though I'm sure some people will blame the product because it allowed an incompatible config. But we'll manage!

sio4 commented:
Hi @stanislas-m,

> https://github.com/gobuffalo/pop/blob/master/pop_test.go already define a suite mechanism, why not using it for your extended tests? :)

Do you mean using `transaction()`? and/or PDB for extra tests of DDLs?
If you mean `transaction()`, I could not use `transaction()` for this since the DDL methods including `CreateDB()` and `DropDB()` was implemented as they create a new connection to the database with *connection details* instead of using existing connection. To prevent other tests were broken, I just use the approach of "getting configured connection" to get working *connection details* and modifying it for DDL tests. I didn't consider about PDB at that time.

Anyway, I cleaned up those test cases slightly simple and more readable, and also added more detailed warning messages for when the user specified those important parameters as options.

Please check this.

Thank you!

stanislas-m commented:
It's not about the transaction, but about the fact we already have a system to manage specific tests for dialects: see https://github.com/gobuffalo/pop/blob/2b6e2a4c979e204a9e61b95ed419780493940079/pop_test.go#L30 and https://github.com/gobuffalo/pop/blob/76471b9d2de258aea1616d3b01b7540f444eea53/slices_test.go#L18 for instance. These tests only run according to the `SODA_DIALECT` env var, so we can test specific behaviors using a real database: this is exactly what you're trying to reimplement with `isExtraTestOnFor`. If you move these tests to the suite (so for MySQL, it will be the MySQLSuite), you'll have the same behavior for free.

I also think we don't need this new `POP_EXTRA_TEST`.

sio4 commented:
Oh, I saw that but I didn't check how it works. It is a similar thing when I found on buffalo app's ActionSuite and ModelSuite! Ok, I will change with it. For `POP_EXTRA_TEST`, I just add it for safety. But since the database names I used on this cases are very strange, it will not serious. I will remove it too!

Anyway, When I use `*Suite` in my apps, it is harder to run a single test from vscode than usual `Test_Something(*testing.T)` since vscode provide a simple link on top of the test case. If you use vscode, is there any hint to make it simple?

stanislas-m commented:
Thanks!

> If you use vscode, is there any hint to make it simple?
I don't know how to do that using VSCode, but maybe you can ask them on their Github repo?
